test/e2e/instrumentation/monitoring/custom_metrics_autoscaling.go

```diff
@@ -0,0 +1,205 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package monitoring
+
+import (
+	"context"
+	"time"
+
+	"golang.org/x/oauth2/google"
+	clientset "k8s.io/client-go/kubernetes"
+
+	. "github.com/onsi/ginkgo"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/test/e2e/framework"
+	instrumentation "k8s.io/kubernetes/test/e2e/instrumentation/common"
+
+	gcm "google.golang.org/api/monitoring/v3"
+	autoscaling "k8s.io/api/autoscaling/v2beta1"
+	"k8s.io/apimachinery/pkg/api/resource"
+)
+
+var _ = instrumentation.SIGDescribe("Stackdriver Monitoring", func() {
+	BeforeEach(func() {
+		framework.SkipUnlessProviderIs("gce", "gke")
+	})
+
+	f := framework.NewDefaultFramework("stackdriver-monitoring")
+	var kubeClient clientset.Interface
+
+	It("should autoscale with Custom Metrics from Stackdriver [Feature:StackdriverMonitoring]", func() {
+		kubeClient = f.ClientSet
+		testHPA(f, kubeClient)
+	})
+})
+
+func testHPA(f *framework.Framework, kubeClient clientset.Interface) {
+	projectId := framework.TestContext.CloudConfig.ProjectID
+
+	ctx := context.Background()
+	client, err := google.DefaultClient(ctx, gcm.CloudPlatformScope)
+
+	// Hack for running tests locally, needed to authenticate in Stackdriver
+	// If this is your use case, create application default credentials:
+	// $ gcloud auth application-default login
+	// and uncomment following lines:
+	/*
+		ts, err := google.DefaultTokenSource(oauth2.NoContext)
+		framework.Logf("Couldn't get application default credentials, %v", err)
+		if err != nil {
+			framework.Failf("Error accessing application default credentials, %v", err)
+		}
+		client := oauth2.NewClient(oauth2.NoContext, ts)
+	*/
+
+	gcmService, err := gcm.New(client)
+	if err != nil {
+		framework.Failf("Failed to create gcm service, %v", err)
+	}
+
+	framework.ExpectNoError(err)
+
+	// Set up a cluster: create a custom metric and set up k8s-sd adapter
+	err = createDescriptors(gcmService, projectId)
+	if err != nil {
+		framework.Failf("Failed to create metric descriptor: %v", err)
+	}
+	defer cleanupDescriptors(gcmService, projectId)
+
+	err = createAdapter()
+	if err != nil {
+		framework.Failf("Failed to set up: %v", err)
+	}
+	defer cleanupAdapter()
+
+	// Run application that exports the metric
+	err = createDeploymentsToScale(kubeClient)
+	if err != nil {
+		framework.Failf("Failed to create stackdriver-exporter pod: %v", err)
+	}
+	defer cleanupDeploymentsToScale(kubeClient)
+
+	// Autoscale the deployments
+	err = createPodsHPA(kubeClient)
+	if err != nil {
+		framework.Failf("Failed to create 'Pods' HPA: %v", err)
+	}
+	err = createObjectHPA(kubeClient)
+	if err != nil {
+		framework.Failf("Failed to create 'Objects' HPA: %v", err)
+	}
+
+	// Wait a for HPA to scale down targets
+	// TODO: wait for the deployments to reach the expected numbers of replicas instead of fixed
+	//       amount of time
+	time.Sleep(240 * time.Second)
```

Comment:
Is there a particular reason not to implement this TODO immediately?

Reply:
Implemented. I left a TODO because I didn't know how much work would it be and this PR was pretty urgent (I must wait for #53330 though)